aleo-programs/staking_lite/src/main.leo

```diff
@@ -0,0 +1,376 @@
+  */
+  mapping state: u8 => u8;
+
+  // 0u8 -> total pool of delegator tokens
+  mapping total_shares: u8 => u64;
+
+  // address -> number of shares held by the delegator with this address
+  mapping delegator_shares: address => u64;
+
+  // 0u8 -> total balance in microcredits, updated for each deposit and withdrawal
+  mapping total_balance: u8 => u64;
+
+  // 0u8 -> balance pending withdrawal in microcredits
+  mapping pending_withdrawal: u8 => u64;
+
+  /** Unbonding allowed: 0u8 ->
+    * The height at which the current withdrawal batch will be done unbonding
+    * if not present or == 0u32, a new batch can begin unbonding
+  */
+  mapping current_batch_height: u8 => u32;
+
+  struct withdrawal_state {
+    microcredits: u64;
+    claim_block: u32;
+  }
+
+  // address -> pending withdrawal for the delegator with this address
+  mapping withdrawals: address => withdrawal_state;
+
+  transition initialize(commission_rate: u128, validator_address: address) {
+    assert_eq(self.caller, ADMIN);
+    assert(commission_rate < PRECISION_UNSIGNED);
+
+    return then finalize(commission_rate, validator_address);
+  }
+
+  finalize initialize(commission_rate: u128, validator_address: address) {
+    commission_percent.set(0u8, commission_rate);
+    validator.set(0u8, validator_address);
+    state.set(0u8, 0u8);
+    total_shares.set(0u8, 0u64);
+    total_balance.set(0u8, 0u64);
+    pending_withdrawal.set(0u8, 0u64);
+    current_batch_height.set(0u8, 0u32);
+  }
+
+  transition initial_deposit(
+    input_record: credits.aleo/credits,
+    microcredits: u64,
+    validator_address: address
+  ) -> credits.aleo/credits {
+    assert_eq(self.caller, ADMIN);
+    // Must be a credits record because credits.aleo uses self.caller for transfers
+    let updated_record: credits.aleo/credits = credits.aleo/transfer_private_to_public(input_record, CORE_PROTOCOL, microcredits);
+    credits.aleo/bond_public(validator_address, microcredits);
+
+    return (updated_record) then finalize(microcredits);
+  }
+
+  finalize initial_deposit(microcredits: u64) {
+    let balance: u64 = total_balance.get_or_use(0u8, 0u64);
+    let shares: u64 = total_shares.get_or_use(0u8, 0u64);
+    assert_eq(balance, 0u64);
+    assert_eq(shares, 0u64);
+
+    total_balance.set(0u8, microcredits);
+    total_shares.set(0u8, microcredits);
+    delegator_shares.set(ADMIN, microcredits);
+  }
+
+  transition set_commission_percent(commission_rate: u128) {
+    assert_eq(self.caller, ADMIN);
+    assert(commission_rate < PRECISION_UNSIGNED);
+
+    return then finalize(commission_rate);
+  }
+
+  finalize set_commission_percent(commission_rate: u128) {
+    let current_state: u8 = state.get(0u8);
+    assert_eq(current_state, 0u8);
+
+    // Make sure all commission is claimed before changing the rate
+    let balance: u64 = total_balance.get(0u8);
+    let bonded: u64 = 0u64; // credits.aleo/bonded.get(CORE_PROTOCOL);
+    assert_eq(balance, bonded);
```

Comment:
I'm not sure this condition can ever realistically be met. Rewards are granted on every block so you'd need to claim rewards within the same block. 

A couple options: 
1. We can have some tolerance for the difference. 
2. We can call `claim_commission` first. 

Although more expensive, I think `claim_commission` makes more sense as finding a correct tolerance is difficult. There's not really a correct option as validators can always jack up the rate because it's configurable.

Reply:
I think we can make changing the commission rate more expensive to guarantee that all the commission is claimed at the previous rate. This will prevent any theft of funds from users adding funds to the staking protocol at a given rate only to have the validator raise commission rates substantially.

I am in favor of option 2.